changefeedccl: Add timeout to testfeed library.

miretskiy opened the pull request:
Add timeout to testfeed sink implementations to timeout
if no messages are received for too long.

Informs #83946

Release Notes: None

Commit:
changefeedccl: Add timeout to testfeed library.

Add timeout to testfeed sink implementations to timeout
if no messages are received for too long.

Release Notes: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/84007)
<!-- Reviewable:end -->

stevendanna commented on pkg/ccl/changefeedccl/testfeed_test.go:
```diff
@@ -127,16 +129,26 @@ type sinklessFeed struct {
 
+func deadline() time.Duration {
+	if util.RaceEnabled {
+		return time.Minute
```
Race can be _really_ slow in CI sometimes. I might even consider something like 5 minutes.

miretskiy replied:
Ack.

stevendanna reviewed (approved):
Looks good. No idea what the appropriate timeout is for Race builds, but it will definitely be nice for these to time out.

stevendanna commented on pkg/ccl/changefeedccl/testfeed_test.go:
```diff
@@ -168,9 +181,9 @@ func (c *sinklessFeed) Next() (*cdctest.TestFeedMessage, error) {
 
 // Resume implements the TestFeed interface.
 func (c *sinklessFeed) start() error {
-	ctx := context.Background()
+	c.ctx, c.cancel = context.WithCancel(context.Background())
```
For sinkless feeds, I think we need a timeout here.  I don't think Query will return until at least one row returns, so if our changefeed never returns any rows, we don't time out.

stevendanna replied:
But we never make it to next.  Because Query never returns:

```
goroutine 38 [IO wait, 1 minutes]:
internal/poll.runtime_pollWait(0x370ba790, 0x72)
        GOROOT/src/runtime/netpoll.go:303 +0x85
internal/poll.(*pollDesc).wait(0xc0006dda00, 0xc000196400, 0x0)
        GOROOT/src/internal/poll/fd_poll_runtime.go:84 +0x32
internal/poll.(*pollDesc).waitRead(...)
        GOROOT/src/internal/poll/fd_poll_runtime.go:89
internal/poll.(*FD).Read(0xc0006dda00, {0xc000196400, 0x8e7, 0x8e7})
        GOROOT/src/internal/poll/fd_unix.go:167 +0x25a
net.(*netFD).Read(0xc0006dda00, {0xc000196400, 0x8e7, 0x0})
        GOROOT/src/net/fd_posix.go:56 +0x29
net.(*conn).Read(0xc001fed320, {0xc000196400, 0xc00601c9f8, 0xc00601ca08})
        GOROOT/src/net/net.go:183 +0x45
crypto/tls.(*atLeastReader).Read(0xc001fcab10, {0xc000196400, 0x0, 0x400ab8d})
        GOROOT/src/crypto/tls/conn.go:778 +0x3d
bytes.(*Buffer).ReadFrom(0xc005fa7e78, {0x9ea9140, 0xc001fcab10})
        GOROOT/src/bytes/buffer.go:204 +0x98
crypto/tls.(*Conn).readFromUntil(0xc005fa7c00, {0x9eb77c0, 0xc001fed320}, 0x4042391)
        GOROOT/src/crypto/tls/conn.go:800 +0xe5
crypto/tls.(*Conn).readRecordOrCCS(0xc005fa7c00, 0x0)
        GOROOT/src/crypto/tls/conn.go:607 +0x112
crypto/tls.(*Conn).readRecord(...)
        GOROOT/src/crypto/tls/conn.go:575
crypto/tls.(*Conn).Read(0xc005fa7c00, {0xc002b4622d, 0x1dd3, 0xc0028d8000})
        GOROOT/src/crypto/tls/conn.go:1278 +0x16f
io.ReadAtLeast({0x13e10e20, 0xc005fa7c00}, {0xc002b4622d, 0x1dd3, 0x1dd3}, 0x5)
        GOROOT/src/io/io.go:328 +0x9a
github.com/jackc/chunkreader/v2.(*ChunkReader).appendAtLeast(...)
        github.com/jackc/chunkreader/v2/external/com_github_jackc_chunkreader_v2/chunkreader.go:88
github.com/jackc/chunkreader/v2.(*ChunkReader).Next(0xc00505a340, 0x5)
        github.com/jackc/chunkreader/v2/external/com_github_jackc_chunkreader_v2/chunkreader.go:78 +0x2ac
github.com/jackc/pgproto3/v2.(*Frontend).Receive(0xc005189000)
        github.com/jackc/pgproto3/v2/external/com_github_jackc_pgproto3_v2/frontend.go:74 +0x50
github.com/jackc/pgconn.(*PgConn).peekMessage(0xc0056cec60)
        github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:499 +0x136
github.com/jackc/pgconn.(*ResultReader).readUntilRowDescription(0xc0056cecf8)
        github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:1529 +0x31
github.com/jackc/pgconn.(*PgConn).execExtendedSuffix(0xc0056cec60, {0xc0028d8000, 0xc004fbce00, 0x0}, 0xc0056cecf8)
        github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:1131 +0x24b
github.com/jackc/pgconn.(*PgConn).ExecPrepared(0xc0056cec60, {0x9f549c0, 0xc004fbce00}, {0xc000567760, 0xa}, {0x0, 0x0, 0x0}, {0x0, 0x0, ...}, ...)
        github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:1076 +0x196
github.com/jackc/pgx/v4.(*Conn).Query(0xc003934870, {0x9f549c0, 0xc004fbce00}, {0x88e5005, 0x19}, {0x0, 0x68, 0x78})
        github.com/jackc/pgx/v4/external/com_github_jackc_pgx_v4/conn.go:651 +0xd8d
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*sinklessFeed).start(0xc0006dd900)
        github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/testfeed_test.go:201 +0x37e
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*sinklessFeedFactory).Feed(0xc00235eb60, {0x88e5005, 0x19}, {0x0, 0x0, 0x0})
        github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/testfeed_test.go:88 +0x19d
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.feed({0xa023850, 0xc00235eb60}, {0x9eaa780, 0xc0040aa780}, {0x88e5005, 0x19}, {0x0, 0x0, 0x0})
        github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:596 +0x98
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.TestChangefeedTestTimesOut.func1(_, {0xc005d5fc70, {0xa055510, 0xc001435dc0}, {{0xc003e15890}, {0xc003e15890}, 0x0}, {{0x0, 0x0}, {0x0, ...}, ...}}, ...)
        github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_test.go:701 +0xe6
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.cdcTestNamed.func1(_, {{0xc005d5fc70, {0xa055510, 0xc001435dc0}, {{0xc003e15890}, {0xc003e15890}, 0x0}, {{0x0, 0x0}, {0x0, ...}, ...}}, ...}, ...)
        github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:820 +0x5d
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.cdcTestNamedWithSystem.func1(0xc00235eb60)
        github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:848 +0x24d
testing.tRunner(0xc00235eb60, 0xc00089fdd0)
        GOROOT/src/testing/testing.go:1259 +0x102
created by testing.(*T).Run
        GOROOT/src/testing/testing.go:1306 +0x35a
```

miretskiy replied:
> But we never make it to next. Because Query never returns:

Ahh.. i see.

miretskiy commented:
Made some additional changes; added the test you wrote -- almost verbatim; seems to correctly
timeout in 100 runs... So, 🤞 ....

miretskiy commented:
cancelling sinkless feed proving to be more problematic; removed WithCancel logic and replaced it with direct call to connection cancellation instead.  Let's see if that makes CI happy.
